File: src/api/projects.ts
import { callCommand } from './client'
import type { Project, ProjectDraft } from '@/types/project'

export function listProjects(): Promise<Project[]> {
  return callCommand<Project[]>('list_projects').then((projects) => projects ?? [])
}

export function createProject(draft: ProjectDraft): Promise<Project> {
  return callCommand<Project>('create_project', { draft })
}

export function updateProject(id: string, draft: ProjectDraft): Promise<Project> {
  return callCommand<Project>('update_project', { id, draft })
}

export function archiveProject(id: string): Promise<void> {
  return callCommand<void>('archive_project', { id })
}

File: src-tauri/src/commands/projects.rs
use tauri::State;

use crate::{
    db::app_state::AppState,
    domain::project::{Project, ProjectDraft, ProjectRepository},
};

#[tauri::command]
pub fn list_projects(state: State<'_, AppState>) -> Result<Vec<Project>, String> {
    state.with_database(|connection| ProjectRepository::new(connection).list_active())
}

#[tauri::command]
pub fn create_project(
    state: State<'_, AppState>,
    draft: ProjectDraft,
) -> Result<Project, String> {
    state.with_database(|connection| ProjectRepository::new(connection).create(draft))
}

#[tauri::command]
pub fn update_project(
    state: State<'_, AppState>,
    id: String,
    draft: ProjectDraft,
) -> Result<Project, String> {
    state.with_database(|connection| ProjectRepository::new(connection).update(&id, draft))
}

#[tauri::command]
pub fn archive_project(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state.with_database(|connection| ProjectRepository::new(connection).archive(&id))
}
